### PR TITLE
Implement collecting errors while tokenizing

### PR DESCRIPTION
**Pull request by Veetaha**
Now we are collecting errors from `rustc_lexer` and returning them in `ParsedToken { token, error }` and `ParsedTokens { tokens, errors }` structures **([UPD]: this is now simplified, see updates bellow)**.

The main changes are introduced in `ra_syntax/parsing/lexer.rs`. It now exposes the following functions and types:

```rust
pub fn tokenize(text: &str) -> ParsedTokens;
pub fn tokenize_append(text: &str, parsed_tokens_to_append_to: &mut ParsedTokens);
pub fn first_token(text: &str) -> Option<ParsedToken>; // allows any number of tokens in text
pub fn single_token(text: &str) -> Option<ParsedToken>; // allows only a single token in text

pub struct ParsedToken  { pub token: Token,       pub error: Option<SyntaxError> }
pub struct ParsedTokens { pub tokens: Vec<Token>, pub errors: Vec<SyntaxError>   }

pub enum TokenizeError { /* Simple enum which reflects rustc_lexer tokenization errors */ }
```
In the first commit I implemented it with iterators, but then decided that since this crate is ad hoc for `rust-analyzer` and we clearly see the places of its usage it would be better to simplify it to vectors.

This is currently WIP, because I want to add tests for error messages generated by the lexer.
I'd like to listen to you thoughts how to define these tests in `ra_syntax/test-data` dir.

Related issues: #223 

**[UPD]**

After the PR review the API was simplified:
```rust
pub fn tokenize(text: &str) -> (Vec<Token>, Vec<SyntaxError>);
// Both lex functions do not check for unescape errors
pub fn lex_single_syntax_kind(text: &str) -> Option<(SyntaxKind, Option<SyntaxError>)>;
pub fn lex_single_valid_syntax_kind(text: &str) -> Option<SyntaxKind>;

// This will be removed in the next PR in favour of simlifying `SyntaxError` to `(String, TextRange)`
pub enum TokenizeError { /* Simple enum which reflects rustc_lexer tokenization errors */ }

// this is private, but may be made public if such demand would exist in future (least privilege principle)
fn lex_first_token(text: &str) -> Option<(Token, Option<SyntaxError>)>;
```

### CODE REVIEW

File: crates/ra_syntax/src/parsing/reparsing.rs [hunk 1]
```diff
@@ -41,37 +41,42 @@ fn reparse_token<'node>(
     root: &'node SyntaxNode,
     edit: &AtomTextEdit,
 ) -> Option<(GreenNode, TextRange)> {
-    let token = algo::find_covering_element(root, edit.delete).as_token()?.clone();
-    match token.kind() {
+    let prev_token = algo::find_covering_element(root, edit.delete).as_token()?.clone();
```
Comment (Veetaha) [inline crates/ra_syntax/src/parsing/reparsing.rs]: Spent some time figuring out what is going on in this method to properly integrate changes from lexer. The core implementation here didn't change, just renamed some variables for clarity...

File: crates/ra_syntax/src/parsing/reparsing.rs [hunk 2]
```diff
@@ -97,6 +102,9 @@ fn reparse_block<'node>(
 fn get_text_after_edit(element: SyntaxElement, edit: &AtomTextEdit) -> String {
     let edit =
         AtomTextEdit::replace(edit.delete - element.text_range().start(), edit.insert.clone());
+
+    // Note: we could move this match to a method or even further: use enum_dispatch crate
+    // https://crates.io/crates/enum_dispatch
```
Comment (Veetaha) [inline crates/ra_syntax/src/parsing/reparsing.rs]: This is my proposal for the future refactoring, want to hear your opinion on that.

File: crates/ra_syntax/src/syntax_error.rs [hunk 3]
```diff
@@ -4,14 +4,18 @@ use std::fmt;
...
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub struct SyntaxError {
     kind: SyntaxErrorKind,
     location: Location,
 }
 
+// FIXME: Location should be just `Location(TextRange)`
+// TextUnit enum member just unnecessarily compicates things,
+// we should'n treat it specially, it just as a `TextRange { start: x, end: x + 1 }`
+// see `location_to_range()` in ra_ide/src/diagnostics
```
Comment (Veetaha) [inline crates/ra_syntax/src/syntax_error.rs]: I think this should be the target for the future refactoring. I am not entirely sure why we don't use just `TextRange` in as a location of `SyntaxError` instead. Or at least make location a newtype of `TextRange`? What do you think?

File: crates/ra_syntax/src/syntax_error.rs [hunk 4]
```diff
@@ -111,6 +120,44 @@ impl fmt::Display for SyntaxErrorKind {
     }
 }
 
+impl fmt::Display for TokenizeError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let msg = match self {
+            TokenizeError::EmptyInt => "Missing digits after integer base prefix",
+            TokenizeError::EmptyExponent => "Missing digits after the exponent symbol",
```
Comment (Veetaha) [inline crates/ra_syntax/src/syntax_error.rs]: I wish `rustfmt` was not so clippy, initialy I wrote this match with each arm homogeneously and nicely wrapped into curly braces, but the formatting tool demands the shorthand syntax here ...

File: crates/ra_syntax/src/syntax_node.rs [hunk 5]
```diff
@@ -38,14 +38,15 @@ pub type SyntaxElementChildren = rowan::SyntaxElementChildren<RustLanguage>;
 
 pub use rowan::{Direction, NodeOrToken};
 
+#[derive(Default)]
 pub struct SyntaxTreeBuilder {
     errors: Vec<SyntaxError>,
     inner: GreenNodeBuilder<'static>,
 }
 
-impl Default for SyntaxTreeBuilder {
```
Comment (Veetaha) [inline crates/ra_syntax/src/syntax_node.rs]: I am not sure why `Default` trait was implemented here by hand, could you, please, elabortate @matklad?

**Comment by Veetaha** [comment]
I know that `SourceFile::parse` returns `Parse<T>` type that has the same semmantics as `ParsedToken[s]`, it could be reasonable to reuse it and make it `Parse<Token>`. I am not sure whether it is worth to generalize it this way...

File: crates/ra_syntax/src/tests.rs [hunk 6]
```diff
@@ -10,7 +10,8 @@ use crate::{fuzz, SourceFile};
 #[test]
 fn lexer_tests() {
     dir_tests(&test_data_dir(), &["lexer"], |text, _| {
-        let tokens = crate::tokenize(text);
+        // FIXME: add tests for errors (their format is up to discussion)
+        let tokens = crate::tokenize(text).tokens;
```
Comment (Veetaha) [inline crates/ra_syntax/src/tests.rs]: Your input is very important here, guys!

File: crates/ra_ide/src/references/rename.rs [hunk 7]
```diff
@@ -17,11 +17,9 @@ pub(crate) fn rename(
...
     new_name: &str,
 ) -> Option<RangeInfo<SourceChange>> {
-    let tokens = tokenize(new_name);
-    if tokens.len() != 1
-        || (tokens[0].kind != SyntaxKind::IDENT && tokens[0].kind != SyntaxKind::UNDERSCORE)
-    {
-        return None;
+    match single_token(new_name)?.token.kind {
+        SyntaxKind::IDENT | SyntaxKind::UNDERSCORE => (),
+        _ => return None,
```
Comment (kiljacken) [inline crates/ra_ide/src/references/rename.rs]: I love how this turned out :+1:

File: crates/ra_syntax/src/parsing/lexer.rs [hunk 8]
```diff
@@ -14,91 +16,242 @@ pub struct Token {
...
+#[derive(Debug, Default)]
+/// Represents the result of parsing source code of Rust language.
+pub struct ParsedTokens {
+    /// Parsed tokens in order they appear in source code.
+    pub tokens: Vec<Token>,
+    /// Collection of all occured tokenization errors.
+    /// In general `self.errors.len() <= self.tokens.len()`
+    pub errors: Vec<SyntaxError>,
+}
+impl ParsedTokens {
```
Comment (kiljacken) [inline crates/ra_syntax/src/parsing/lexer.rs]: Nit: Add a blank line above this.

File: crates/ra_syntax/src/parsing/lexer.rs [hunk 9]
```diff
@@ -14,91 +16,242 @@ pub struct Token {
...
     };
-    Some(Token { kind, len: TextUnit::from_usize(t.len) })
+
+    return ParsedToken {
+        token: Token { kind: syntax_kind, len: token_range.len() },
+        error: error
+            .map(|error| SyntaxError::new(SyntaxErrorKind::TokenizeError(error), token_range)),
+    };
+
+    type ParsedSyntaxKind = (SyntaxKind, Option<TokenizeError>);
```
Comment (kiljacken) [inline crates/ra_syntax/src/parsing/lexer.rs]: Defining the alias here probably takes up more space that just writing it out four times. Don't really care too much either way.

File: crates/ra_syntax/src/parsing/reparsing.rs (hunk 1 above)
Comment (kiljacken) [reply crates/ra_syntax/src/parsing/reparsing.rs]: Good names do a lot for code readability :+1:

File: crates/ra_syntax/src/parsing/reparsing.rs (hunk 2 above)
Comment (kiljacken) [reply crates/ra_syntax/src/parsing/reparsing.rs]: I think there's some consensus that we already have a bit too many dependencies, so adding another one, especially a procedural macro, is probably not a good thing if we want to keep build times down.

File: crates/ra_syntax/src/parsing/reparsing.rs [hunk 10]
```diff
@@ -182,6 +193,8 @@ mod tests {
         let fully_reparsed = SourceFile::parse(&after);
         let incrementally_reparsed: Parse<SourceFile> = {
             let f = SourceFile::parse(&before);
+            // FIXME: it seems this initialization statement is unnecessary (see edit in outer scope)
+            // Investigate whether it should really be removed.
             let edit = AtomTextEdit { delete: range, insert: replace_with.to_string() };
```
Comment (kiljacken) [inline crates/ra_syntax/src/parsing/reparsing.rs]: Neither `range` nor `replace_with` change, and `edit` is always passed as reference, so I'm inclined to agree that this line could probably be removed.

File: crates/ra_syntax/src/syntax_error.rs (hunk 3 above)
Comment (kiljacken) [reply crates/ra_syntax/src/syntax_error.rs]: Seems reasonable. In the end, when used to emit a diagnostic, it gets turned into a TextRange anyways: https://github.com/rust-analyzer/rust-analyzer/blob/d1330a4a65f0113c687716a5a679239af4df9c11/crates/ra_ide/src/diagnostics.rs#L118-L123

File: crates/ra_syntax/src/syntax_error.rs (hunk 4 above)
Comment (kiljacken) [reply crates/ra_syntax/src/syntax_error.rs]: You could always `#[rustfmt::skip]` this one function. We do the same for a few match statements elsewhere.

File: crates/ra_syntax/src/syntax_error.rs [hunk 11]
```diff
@@ -79,6 +83,10 @@ impl fmt::Display for SyntaxError {
 pub enum SyntaxErrorKind {
     ParseError(ParseError),
     EscapeError(EscapeError),
+    TokenizeError(TokenizeError),
+    // FIXME: the obvious pattern of this enum dictates that the following enum variants
+    // should be wrapped into something like `SemmanticError(SemmanticError)`
+    // or `ValidateError(ValidateError)` or `SemmanticValidateError(...)`
```
Comment (kiljacken) [inline crates/ra_syntax/src/syntax_error.rs]: Lets leave it at the tokenizer changes, at least for this PR. Then you can do this as a separate PR if it seems useful.

File: crates/ra_syntax/src/tests.rs (hunk 6 above)
Comment (kiljacken) [reply crates/ra_syntax/src/tests.rs]: You could probably just update [`dump_tokens`](https://github.com/rust-analyzer/rust-analyzer/blob/d1330a4a65f0113c687716a5a679239af4df9c11/crates/ra_syntax/src/tests.rs#L104-L115) to take `ParsedTokens` and then also output the errors like we do the tokens. If you don't print anything when there's no errors, you shouldn't even need to touch the existing tests.

**Comment by kiljacken** [comment]
> I know that `ra_syntax::SourceFile::parse()` returns `Parse<SourceFile>` type that has the same semmantics as `ParsedToken[s]`, we could reuse it (witch some changes to it) and return `Parse<Token>` instead. But this would be the task for refactoring and I am not sure whether it is worth to generalize it this way... What are your thoughts?

`Parse<T>` is currently rather specialized for things that actually contain a syntax tree from rowan, so this should probably just remain separate.

File: crates/ra_syntax/src/syntax_error.rs (hunk 3 above)
Comment (Veetaha) [reply crates/ra_syntax/src/syntax_error.rs]: Yeah, I just referred to that snippet in the comment)

Comment (kiljacken) [reply crates/ra_syntax/src/syntax_error.rs]: That you did :laughing:

File: crates/ra_syntax/src/syntax_node.rs (hunk 5 above)
Comment (matklad) [reply crates/ra_syntax/src/syntax_node.rs]: At the time this was written, I forgot to add a `Default` to `GreenBuilder`, and was to lazy to publish a new version of rowan just for that: https://github.com/rust-analyzer/rowan/commit/a3692a9a5ed12f8eff5bb516860b9f2fd2495b6d#diff-1a2b23ceb6e534bd16f63d2e55e537aaR154.

In general, if something looks like it doesn't make sense, it probably indeed doesn't make sense :)

File: crates/ra_syntax/src/tests.rs (hunk 6 above)
Comment (matklad) [reply crates/ra_syntax/src/tests.rs]: Just dumping errors in `dump_tokens` after the tokens themselves seems good to me.

Unrelated, but perhaps a  slightly more information rich way to write this would be:

```rust
let ParsedToken { tokens, errors: _ } = crate::tokenize(text)
```

Ie, explicitely naming the thing you are ignoring (which usually is important for erroes).

File: crates/ra_syntax/src/syntax_error.rs [hunk 12]
```diff
@@ -126,31 +126,31 @@ impl fmt::Display for TokenizeError {
...
-                "Missing trailing `\"` with `#` symbols to terminate the raw string literal"
+                "Missing trailing \" with # symbols to terminate the raw string literal"
             }
             TokenizeError::UnterminatedRawByteString => {
-                "Missing trailing `\"` with `#` symbols to terminate the raw byte string literal"
+                "Missing trailing \" with # symbols to terminate the raw byte string literal"
             }
             TokenizeError::UnstartedRawString => {
-                "Missing `\"` symbol after `#` symbols to begin the raw string literal"
+                "Missing \" symbol after # symbols to begin the raw string literal"
```
Comment (matklad) [inline crates/ra_syntax/src/syntax_error.rs]: I usually like to place some kind of quotes around snippets because that's helpful when a snippet contains trailing/leading whitespace or other invisible characters. But we should just stick to what rustc is doing.

Which... I am not sure what is exactly! For this code

```rust
λ cat main.rs 
fn main() {
    let s = r###;
}
```

I get this error

image

Note how `#` is uselessly in quotes, but the empty string between `:` and  `;` isn't.

@estebank what are the current guidelines around quotes?

File: crates/ra_syntax/src/syntax_error.rs (hunk 11 above)
Comment (Veetaha) [reply crates/ra_syntax/src/syntax_error.rs]: Yes, I described the summary and future todos in the comment [under the original issue](https://github.com/rust-analyzer/rust-analyzer/issues/223#issuecomment-578559627). I'll see what should be done in next PR.

File: crates/ra_syntax/src/parsing/reparsing.rs (hunk 2 above)
Comment (Veetaha) [reply crates/ra_syntax/src/parsing/reparsing.rs]: Okay, guys, that you for the opinion)

File: crates/ra_syntax/src/syntax_node.rs (hunk 5 above)
Comment (Veetaha) [reply crates/ra_syntax/src/syntax_node.rs]: I've stumbled with enough amount of things that didn't make sense to me but did actually exist for a reason, so I'd rather ask the author first before amending them. Anyway, thank you for the clarification!

File: crates/ra_syntax/src/tests.rs (hunk 6 above)
Comment (Veetaha) [reply crates/ra_syntax/src/tests.rs]: Okay guys, will do!
Also, @matklad, good point on explicit errors dropping. Do you think we should use this pattern in all places `tokenize*()/single_token()/first_token()` are called?

Comment (matklad) [reply crates/ra_syntax/src/tests.rs]: I think so! In general, I just always try to destruct tuples that way:

```
13:49:36|~/projects/rust-analyzer|HEAD✓
λ rg  'let .*, _\w+\)'
xtask/src/codegen/gen_syntax.rs
167:    let punctuation_values = grammar.punct.iter().map(|(token, _name)| {

crates/ra_hir_def/src/generics.rs
59:        let (params, _source_map) = GenericParams::new(db, def.into());

crates/ra_hir/src/code_model.rs
1072:        let (adt, _subst) = self.ty.value.as_adt()?;

crates/ra_batch/src/lib.rs
53:    let (crate_graph, _crate_names) =
147:        let (host, _roots) = load_cargo(path).unwrap();

crates/ra_hir_ty/src/utils.rs
141:            let (_total, parent_len, _child) = self.len_split();

crates/ra_parser/src/grammar/expressions.rs
27:    let (cm, _block_like) = expr(p);

crates/ra_hir_def/src/nameres/collector.rs
947:        let (db, _file_id) = TestDB::with_single_file(&code);

crates/ra_hir_ty/src/infer/coerce.rs
253:                let (last_field_id, _data) = fields.next_back()?;

crates/ra_parser/src/grammar/expressions/atom.rs
560:    let (completed, _is_block) =
``

File: crates/ra_ide/src/references/rename.rs [hunk 13]
```diff
@@ -17,11 +17,9 @@ pub(crate) fn rename(
     position: FilePosition,
     new_name: &str,
 ) -> Option<RangeInfo<SourceChange>> {
-    let tokens = tokenize(new_name);
-    if tokens.len() != 1
-        || (tokens[0].kind != SyntaxKind::IDENT && tokens[0].kind != SyntaxKind::UNDERSCORE)
-    {
-        return None;
+    match single_token(new_name)?.token.kind {
```
[the remark this answers is not in the document]
Comment (Veetaha) [reply crates/ra_ide/src/references/rename.rs]: Hmm, I know that `Parse<T>` has `ok()` method. Maybe `ParsedToken` should declare the same?

https://github.com/rust-analyzer/rust-analyzer/blob/a108f22d835c8d67c93c66758d89371fff179b9a/crates/ra_syntax/src/lib.rs#L95-L102

File: crates/ra_syntax/src/parsing.rs [hunk 14]
```diff
@@ -7,15 +7,17 @@ mod text_tree_sink;
...
 
 pub(crate) use self::reparsing::incremental_reparse;
 
 pub(crate) fn parse_text(text: &str) -> (GreenNode, Vec<SyntaxError>) {
-    let tokens = tokenize(&text);
-    let mut token_source = text_token_source::TextTokenSource::new(text, &tokens);
-    let mut tree_sink = text_tree_sink::TextTreeSink::new(text, &tokens);
+    let ParsedTokens { tokens, errors } = tokenize(&text);
+    let mut token_source = TextTokenSource::new(text, &tokens);
+    let mut tree_sink = TextTreeSink::new(text, &tokens, errors);
```
[the remark this answers is not in the document]
Comment (Veetaha) [reply crates/ra_syntax/src/parsing.rs]: Hmm, it seemed to me that since parsing happens very frequently we would like to avoid additional allocations here, thus I passed the resulting vector directly to the constructor. If this is not critical, I can change it to merging errors. Well, are you sure, @matklad ?

[the remark this answers is not in the document]
Comment (matklad) [reply crates/ra_syntax/src/parsing.rs]: I'd prioritize readability over performance, unless perf improvements come with benchmarks. In this particular case, I think most of the time both vectors are empty, so this shouldn't matter much. Even if they are not empty, the cost of an memcpy and allocation would probably be dwarfed by the cost of allocating the syntax tree itself, which does an allocation per node

File: crates/ra_syntax/src/parsing/lexer.rs [hunk 15]
```diff
@@ -13,85 +15,242 @@ pub struct Token {
...
+        TK::Minus => ok(MINUS),
+        TK::And => ok(AMP),
+        TK::Or => ok(PIPE),
+        TK::Plus => ok(PLUS),
+        TK::Star => ok(STAR),
+        TK::Slash => ok(SLASH),
+        TK::Caret => ok(CARET),
+        TK::Percent => ok(PERCENT),
+        TK::Unknown => ok(ERROR),
+    };
```
[the remark this answers is not in the document]
Comment (Veetaha) [reply crates/ra_syntax/src/parsing/lexer.rs]: Hmm, nice notice, I'll apply this

File: crates/ra_syntax/src/syntax_error.rs [hunk 16]
```diff
@@ -84,6 +84,9 @@ pub enum SyntaxErrorKind {
     ParseError(ParseError),
     EscapeError(EscapeError),
     TokenizeError(TokenizeError),
+    // FIXME: the obvious pattern of this enum dictates that the following enum variants
+    // should be wrapped into something like `SemmanticError(SemmanticError)`
+    // or `ValidateError(ValidateError)` or `SemmanticValidateError(...)`
```
[the remark this answers is not in the document]
Comment (Veetaha) [reply crates/ra_syntax/src/syntax_error.rs]: Okay, let's leave it for the refactoring to be done soon. Since this crate is called `ra_syntax` let's leave `SyntaxError` name for that and not some other `OpaqueError`.
If this crate is used only be rust-analyzer exclusively I am definitely for generic `(String, TextRange)` error shape.

File: crates/ra_syntax/src/parsing.rs [hunk 17]
```diff
@@ -7,15 +7,17 @@ mod text_tree_sink;
...
-pub use self::lexer::{classify_literal, tokenize, Token};
+pub use lexer::*;
 
 pub(crate) use self::reparsing::incremental_reparse;
 
 pub(crate) fn parse_text(text: &str) -> (GreenNode, Vec<SyntaxError>) {
-    let tokens = tokenize(&text);
-    let mut token_source = text_token_source::TextTokenSource::new(text, &tokens);
-    let mut tree_sink = text_tree_sink::TextTreeSink::new(text, &tokens);
+    let ParsedTokens { tokens, errors } = tokenize(&text);
```
[the remark this answers is not in the document]
Comment (Veetaha) [reply crates/ra_syntax/src/parsing.rs]: Heh, I see you a great lover of tuples) Maybe, in this case, it is justified since `ParsedToken[s]` is really just an [anemic struct](https://en.wikipedia.org/wiki/Anemic_domain_model).

File: crates/ra_syntax/src/parsing.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (Veetaha) [reply crates/ra_syntax/src/parsing.rs]: Hmm, you are right, this way it seems reasonable to me, thanks!

File: crates/ra_syntax/src/parsing/lexer.rs (hunk 9 above)
Comment (Veetaha) [reply crates/ra_syntax/src/parsing/lexer.rs]: Okay, I'll try to refactor it according to [this comment](https://github.com/rust-analyzer/rust-analyzer/pull/2911/files#r371165828)

File: crates/ra_syntax/src/syntax_error.rs (hunk 12 above)
Comment (Veetaha) [reply crates/ra_syntax/src/syntax_error.rs]: Heh, I initially wrapped the symbols into backticks (hoping they would be rendered as Markdown), but once I started VSCode and checked out that the messages were displayed as a raw text I removed them in this commit:
https://github.com/rust-analyzer/rust-analyzer/pull/2911/commits/34d72aa0c9bb73a2dbfa3ed6dedd5fe88e4aaed7

File: crates/ra_syntax/src/parsing/lexer.rs [hunk 18]
```diff
@@ -13,85 +15,242 @@ pub struct Token {
...
+#[derive(Debug)]
+/// Represents the result of parsing one token.
+pub struct ParsedToken {
+    /// Parsed token.
+    pub token: Token,
+    /// If error is present then parsed token is malformed.
+    pub error: Option<TokenizeError>,
+}
+impl ParsedToken {
+    pub const fn new(token: Token, error: Option<TokenizeError>) -> Self {
```
[the remark this answers is not in the document]
Comment (Veetaha) [reply crates/ra_syntax/src/parsing/lexer.rs]: Didn't expect you would dig into commits, I did rethink this and removed that functions in later commits.

File: crates/ra_syntax/src/parsing/lexer.rs [hunk 19]
```diff
@@ -13,85 +15,242 @@ pub struct Token {
...
+    rustc_token_kind: &rustc_lexer::TokenKind,
+    token_text: &str,
+) -> ParsedToken {
+    use rustc_lexer::TokenKind as TK;
+    use TokenizeError as TE;
+
+    // We drop some useful infromation here (see patterns with double dots `..`)
+    // Storing that info in `SyntaxKind` is not possible due to its layout requirements of
+    // being `u16` that come from `rowan::SyntaxKind` type and changes to `rowan::SyntaxKind`
+    // would mean hell of a rewrite.
```
[the remark this answers is not in the document]
Comment (Veetaha) [reply crates/ra_syntax/src/parsing/lexer.rs]: Sorry, I think I did make this comment a bit misleading, I just wanted to explain why we do drop some info here, not proposing to change the design!

File: crates/ra_ide/src/references/rename.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (Veetaha) [reply crates/ra_ide/src/references/rename.rs]: Well, since the tuple pattern has become a common practice here, lets leave it that way instead of `ParsedToken[s]`

**Comment by Veetaha** [comment]
Thank you for the review and elaborating on the "rust-analyzer" design, will take all that into account!

File: crates/ra_ide/src/references/rename.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (Veetaha) [reply crates/ra_ide/src/references/rename.rs]: @matklad, `lex_single_token()` should return `Token` but not `SyntaxKind`, shouldn't it?
What about `try_single_token() -> Option<Token>` instead? So this would look like:
```suggestion
    match try_single_token(new_name)?.kind {
```

[the remark this answers is not in the document]
Comment (Veetaha) [reply crates/ra_ide/src/references/rename.rs]: Also, is it conventional to omit `get`, `parse`, `fetch` or other synonyms in function names?

**Comment by matklad** [comment]
is this still a wip?

**Comment by Veetaha** [comment]
@matklad please don't merge yet, I'll add tests today. Then I'll remove WIP label. Or do you have another understanding of WIP label?

**Comment by matklad** [comment]
Sometimes wip lables are left there accidentally, just checking that that's
not the case :)

**Comment by Veetaha** [comment]
@kiljacken, @matklad 
Please see changes of the latest commit: 38f1abfea90c38e609bbde887e13eab5f95d1f78
Brief changelog:
* Extracted `ok` and `err` directories in `test_data/lexer` similarly to how it is done in `test_data/parser`.
* Moved a bunch of existing test for error scenarios to `err` directory, renamed `comments.rs` test file to `single_line_comments.rs`
* Added a whole bunch of tests for tokenization errors introduced in this PR
* Added `FIXME`s (which are in fact todos) to state the plan of future improvements.

**Comment by Veetaha** [comment]
I also need to rebase this branch, I'll do it before the merge, just let me know when it is time.

File: crates/ra_syntax/src/tests.rs [hunk 20]
```diff
@@ -1,18 +1,28 @@
...
+    dir_tests(&test_data_dir(), &["lexer/ok"], |text, path| {
+        let (tokens, errors) = tokenize(text);
+        assert_errors_are_absent(&errors, path);
+        dump_tokens_and_errors(&tokens, &errors, text)
+    });
+    dir_tests(&test_data_dir(), &["lexer/err"], |text, path| {
+        let (tokens, errors) = tokenize(text);
+        assert_errors_are_present(&errors, path);
+        dump_tokens_and_errors(&tokens, &errors, text)
+    });
```
Comment (matklad) [inline crates/ra_syntax/src/tests.rs]: :+1:

**Comment by kiljacken** [comment]
Bors is stuck, we probably need that rebase.

@Veetaha Would you mind rebasing? :)

**Comment by Veetaha** [comment]
Sure, I'll do that later today!

File: crates/ra_syntax/src/syntax_error.rs (hunk 12 above)
Comment (estebank) [reply crates/ra_syntax/src/syntax_error.rs]: We try to consistently surround code and tokens (including identifiers) in backticks. No tool is actually _parsing_ them, but having them bare makes it harder to read (consider <pre>found , but expected .</pre> vs <pre>found \`,\` but expected \`.\`</pre>)

**Comment by Veetaha** [comment]
@matklad bors r+ ?
